### PR TITLE
Implements route backend timeout

### PR DESCRIPTION
**Pull request by AlexanderYastrebov**
* Adds `backendTimeout` filter to configure route backend timeout
* Proxy sets up request context with configured timeout and responds
with 504 status on timeout (note: if response streaming has already
started it will be terminated, client will receive backend status and
truncated response body).

See #1041

### CODE REVIEW

**Comment by AlexanderYastrebov** [comment]
If response truncation is not acceptable I see three ways out:
* declare it a feature for limited usecases :)
* try to workaround/disable streaming termination
* attempt another approach - consider timeout as `Transport.ResponseHeaderTimeout`, keep a pool of transports per configured timeout (cleanup?) and wrap roundtripper around here https://github.com/zalando/skipper/blob/8f32650afa693de7642e87026af0c6f39af179e9/proxy/proxy.go#L951

**Comment by AlexanderYastrebov** [comment]
There is a general problem to notify client about backend that has stuck. At the moment `copyStream` stops writing on backend read error: 
https://github.com/zalando/skipper/blob/8f32650afa693de7642e87026af0c6f39af179e9/proxy/proxy.go#L411-L421
which leads to truncated response. 

`httputil.ReverseProxy` uses [`panic(http.ErrAbortHandler)`](https://github.com/golang/go/blob/d75b7f9648bb4968a5d980c4a4b7aa6ca22c4de5/src/net/http/httputil/reverseproxy.go#L328-L339), that [closes client connection](https://github.com/golang/go/blob/d75b7f9648bb4968a5d980c4a4b7aa6ca22c4de5/src/net/http/server.go#L1813-L1824), see https://github.com/golang/go/issues/23643 (for http/2 it seems to [close all client streams](https://github.com/golang/go/blob/d75b7f9648bb4968a5d980c4a4b7aa6ca22c4de5/src/net/http/h2_bundle.go#L7405-L7411))

**Comment by aryszka** [comment]
> If response truncation is not acceptable I see three ways out:
> ...
> try to workaround/disable streaming termination

I believe this is possible if we implement our variant of context.Context and signal to it various phases like request headers sent, request stream done, response headers received, response stream done. But in my opinion, the current implementation is a good start, and covers a reasonable majority of the use cases.

File: filters/builtin/timeout.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+		tf.timeout = v
+	default:
+		return nil, filters.ErrInvalidFilterParameters
+	}
+	return &tf, nil
+}
+
+func (t *timeout) Request(ctx filters.FilterContext) {
+	sb := ctx.StateBag()
+	if _, ok := sb[filters.BackendTimeout]; !ok { // once
```
Comment (aryszka) [inline filters/builtin/timeout.go:42]: I would allow overriding. Eg it may make sense when the a default timeout is set via the global default filters. Just drop the if.

File: docs/reference/filters.md [hunk 2]
```diff
@@ -466,6 +466,23 @@ Example:
 * -> randomContent(42) -> <shunt>;
 ```
 
+## backendTimeout
+
+Configure backend timeout. Skipper responds with `504 Gateway Timeout` status if obtaining a connection,
+sending the request, and reading the backend response headers and body takes longer than the configured timeout.
+However, if response streaming has already started it will be terminated, i.e. client will receive backend response
+status and truncated response body.
```
Comment (aryszka) [inline docs/reference/filters.md:474]: I think we should include here that if the request body is received too slowly, that also cancels the flow.

**Comment by aryszka** [comment]
I would allow multiple filters where the last one would take effect (see my comment for the `timeout.Request` method. Apart of that, lgtm

File: docs/reference/filters.md (hunk 2 above)
Comment (AlexanderYastrebov) [reply docs/reference/filters.md:474]: Doesn't `sending the request` covers it?

File: filters/builtin/timeout.go (hunk 1 above)
Comment (AlexanderYastrebov) [reply filters/builtin/timeout.go]: Done

File: proxy/backendtimeout_test.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	}
+}
+
+func TestFastService(t *testing.T) {
+	service := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(2 * time.Millisecond)
+	}))
+	defer service.Close()
+
+	doc := fmt.Sprintf(`* -> backendTimeout("10ms") -> "%s"`, service.URL)
```
Comment (szuecs) [inline proxy/backendtimeout_test.go:44]: this might always fail in CI tests

File: proxy/backendtimeout_test.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+	"time"
+)
+
+func TestSlowService(t *testing.T) {
+	service := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(2 * time.Millisecond)
+	}))
+	defer service.Close()
+
+	doc := fmt.Sprintf(`* -> backendTimeout("1ms") -> "%s"`, service.URL)
```
Comment (szuecs) [inline proxy/backendtimeout_test.go:18]: this might always be ok in CI tests

File: proxy/proxy.go [hunk 5]
```diff
@@ -1237,6 +1246,7 @@ func (p *Proxy) errorResponse(ctx *context, err error) {
 	}
 
 	if span := ot.SpanFromContext(ctx.Request().Context()); span != nil {
+		p.tracing.setTag(span, ErrorTag, true)
```
Comment (szuecs) [inline proxy/proxy.go:1249]: this is an unrelated change, which can hurt some of our production teams

Comment (szuecs) [reply proxy/proxy.go:1249]: ah got it, was moved from ServeHTTP

**Comment by szuecs** [comment]
Code looks good to me.
I am pretty sure tests will hurt in the future (1ms/2ms sleeps to separate cases will break too often)

File: docs/reference/filters.md (hunk 2 above)
Comment (aryszka) [reply docs/reference/filters.md:493]: true, sgtm

File: proxy/backendtimeout_test.go (hunk 4 above)
Comment (aryszka) [reply proxy/backendtimeout_test.go:18]: one way to test this better could be a service that never responds:

```
<-testDone
```

File: proxy/backendtimeout_test.go (hunk 3 above)
Comment (aryszka) [reply proxy/backendtimeout_test.go:44]: here we can set a higher TO, e.g. 1s. I don't see a really good way to test this in a more "scientifically" exact way.

**Comment by AlexanderYastrebov** [comment]
I have rebased the PR. I have to think about tests, at the moment I don't quite see how it will hurt. But I can assure you I don't like flaky tests at all.

File: proxy/backendtimeout_test.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+)
+
+func TestSlowService(t *testing.T) {
+	service := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(10 * time.Millisecond)
```
Comment (szuecs) [inline proxy/backendtimeout_test.go:14]: replace with:
<-testDone

File: proxy/backendtimeout_test.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+)
+
+func TestSlowService(t *testing.T) {
```
Comment (szuecs) [inline proxy/backendtimeout_test.go:12]: create testDone channel:
testDone := make(chan struct{})

File: proxy/backendtimeout_test.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+	defer ps.Close()
+
+	rsp, err := http.Get(ps.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if rsp.StatusCode != http.StatusGatewayTimeout {
+		t.Errorf("expected 504, got: %v", rsp)
+	}
```
Comment (szuecs) [inline proxy/backendtimeout_test.go:35]: close(testChan)

File: proxy/backendtimeout_test.go (hunk 3 above)
Comment (szuecs) [reply proxy/backendtimeout_test.go:44]: true let's keep like this and see

File: proxy/backendtimeout_test.go (hunk 7 above)
Comment (AlexanderYastrebov) [reply proxy/backendtimeout_test.go:12]: Fixed

File: proxy/proxy.go [hunk 9]
```diff
@@ -924,10 +927,16 @@ func (p *Proxy) makeBackendRequest(ctx *context) (*http.Response, *proxyError) {
...
 
 		// Check if the request has been cancelled or timed out
-		// The roundtrip error `err` may be different
+		// The roundtrip error `err` may be different:
+		// - for `Canceled` it could be either the same `context canceled` or `unexpected EOF` (net.OpError)
+		// - for `DeadlineExceeded` it is net.Error(timeout=true, temporary=true) wrapping this `context deadline exceeded`
 		if cerr := req.Context().Err(); cerr != nil {
 			ctx.proxySpan.LogKV("event", "error", "message", cerr.Error())
-			return nil, &proxyError{err: cerr, code: 499}
+			if cerr == stdlibcontext.Canceled {
```
Comment (AlexanderYastrebov) [inline proxy/proxy.go:935]: This now covers client cancellation and backend request timeout

File: proxy/proxy.go [hunk 10]
```diff
@@ -1153,6 +1167,7 @@ func (p *Proxy) serveResponse(ctx *context) {
 	copyHeader(ctx.responseWriter.Header(), ctx.response.Header)
 	p.tracing.logStreamEvent(ctx.proxySpan, StreamHeadersEvent, EndEvent)
 
+	// TODO differentiate client cancelation and backend timeout before streaming
```
Comment (AlexanderYastrebov) [inline proxy/proxy.go:1170]: Needs to be implemented and test covered, see similar https://github.com/zalando/skipper/pull/1566#discussion_r538839840

Comment (AlexanderYastrebov) [reply proxy/proxy.go:1170]: Added the test for client cancellation. This codepath is not a subject of backendTimeout as it configures timeout on backend request, not on client request.

**Comment by AlexanderYastrebov** [comment]
I have rebased and updated tests as suggested

File: proxy/proxy.go (hunk 9 above)
Comment (szuecs) [reply proxy/proxy.go:935]: nice!

**Comment by szuecs** [comment]
:+1:

**Comment by aryszka** [comment]
Thanks a lot!

**Comment by aryszka** [comment]
:+1:
